README.md
# MyLab.HttpMetrics
[![NuGet Version and Downloads count](https://buildstats.info/nuget/MyLab.HttpMetrics)](https://www.nuget.org/packages/MyLab.HttpMetrics)

```
Поддерживаемые платформы: .NET Core 3.1+
```
Ознакомьтесь с последними изменениями в [журнале изменений](/changelog.md).

## Обзор

Использует стандартную библиотеку [Prometheus для .NET Core](https://github.com/prometheus-net/prometheus-net) для предоставления основных метрик, связанных с выполнением HTTP запросов.

Для добавления метрик при конфигурировании приложения необходимо с помощью методов расширения:

1. `AddUrlBasedHttpMetrics` - добавить необходимые зависимости 
2. `UseUrlBasedHttpMetrics` - добавить `middleware` для перехвата и анализа запросов 

```C#
public class Startup
{
    public void ConfigureServices(IServiceCollection services)
    {
        services.AddControllers();
        services.AddUrlBasedHttpMetrics();		// <--- 1
    }

    public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
    {
        if (env.IsDevelopment())
        {
            app.UseDeveloperExceptionPage();
        }

        app.UseRouting();

        app.UseUrlBasedHttpMetrics();			// <--- 2

        app.UseEndpoints(endpoints =>
		{
        	endpoints.MapControllers();
            endpoints.MapMetrics();
        });
    }
}
```

Пример вывода собираемых метрик (подрезано):

```
# HELP ml_http_request_count_total The total number of requests
# TYPE ml_http_request_count_total counter
ml_http_request_count_total{method="GET",path="/get/xxx",status_code="200"} 1
# HELP ml_http_request_content_size_bytes The request content size in bytes.
# TYPE ml_http_request_content_size_bytes histogram
ml_http_request_content_size_bytes_sum{method="GET",path="/get/xxx",status_code="200"} 0
ml_http_request_content_size_bytes_count{method="GET",path="/get/xxx",status_code="200"} 1
ml_http_request_content_size_bytes_bucket{method="GET",path="/get/xxx",status_code="200",le="1024"} 1
...
ml_http_request_content_size_bytes_bucket{method="GET",path="/get/xxx",status_code="200",le="+Inf"} 1
# HELP ml_http_request_size_bytes_total The total size of request content
# TYPE ml_http_request_size_bytes_total counter
# HELP ml_http_request_duration_seconds The duration in seconds between the response to a request.
# TYPE ml_http_request_duration_seconds histogram
ml_http_request_duration_seconds_sum{method="GET",path="/get/xxx",status_code="200"} 0.0303877
ml_http_request_duration_seconds_count{method="GET",path="/get/xxx",status_code="200"} 1
ml_http_request_duration_seconds_bucket{method="GET",path="/get/xxx",status_code="200",le="0.01"} 0
...
ml_http_request_duration_seconds_bucket{method="GET",path="/get/xxx",status_code="200",le="30"} 1
ml_http_request_duration_seconds_bucket{method="GET",path="/get/xxx",status_code="200",le="+Inf"} 1
# HELP process_working_set_bytes Process working set
# TYPE process_working_set_bytes gauge
process_working_set_bytes 111038464
# HELP ml_http_response_content_size_bytes The response content size in bytes.
# TYPE ml_http_response_content_size_bytes histogram
ml_http_response_content_size_bytes_sum{method="GET",path="/get/xxx",status_code="200"} 0
ml_http_response_content_size_bytes_count{method="GET",path="/get/xxx",status_code="200"} 1
ml_http_response_content_size_bytes_bucket{method="GET",path="/get/xxx",status_code="200",le="1024"} 1
...
ml_http_response_content_size_bytes_bucket{method="GET",path="/get/xxx",status_code="200",le="+Inf"} 1
```

## Метрики

### `ml_http_request_count_total`

Счётчик. Считает количество поступивших запросов.

### `ml_http_unhandled_exception_total`

Счётчик. Считает количество необработанных исключений.

### `ml_http_request_content_size_bytes`

Гистограмма. Ведёт учёт объёма запросов в байтах. Учёт основан на значениях из заголовка `Content-Length`.

Счётчики гистограммы:

* 1kb
* 2kb                
* 5kb       
* 10kb       
* 20kb       
* 50kb                
* 1Mb

### `ml_http_response_content_size_bytes` 

Гистограмма. Ведёт учёт объёма ответов в байтах. Учёт основан на количестве данных, записанных в выходной поток тела сообщения ответа.

Счётчики гистограммы:

* 1kb
* 2kb                
* 5kb       
* 10kb       
* 20kb       
* 50kb                
* 1Mb

### `ml_http_request_duration_seconds`

Гистограмма. Ведёт учёт длительности обработки запросов в секундах. 

Счётчики гистограммы:

- 10ms
- 20ms
- 50ms
- 100ms
- 200ms
- 500ms
- 1s
- 2s
- 3s
- 10s
- 20s
- 30s

### Метки 

* `method` - HTTP-метод в верхнем регистре
* `path` - нормализованный путь из `URL` запроса
* `status_code` - числовой статус-код ответа

### Нормализация пути URL

Основной целью нормализации является приведение меток метрик, содержащих пути URL запросов, содержащих идентификаторы, к единому виду.

Выполняются следующие манипуляции:

* адрес приводится к нижнему регистру
* проверяется каждый элемент пути и заменяется на `xxx`, если:
  * цифр больше остальных символов
  * это `GUID`
  * содержит последовательность из цифр длиной 3 и более

CHANGELOG.md
# Лог изменений

Все заметные изменения в этом проекте будут отражаться в этом документе.

Формат лога изменений базируется на [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## [1.1.3] - 2021-06-22

### Изменено:

* Переход на `MyLab.Log.Dsl`

## [1.1.2] - 2021-04-30

### Добавлено:

* Конвертор перечислений для использования в метриках `EnumConverter`.

## [1.0.2] - 2020-08-27

### Исправлено:

* Игнорировались исключения при выполнении запроса и выдавался ответ `200`.

## [1.0.1] - 2020-08-04

### Удалено:

* Удалены лишние метрики `ml_http_request_size_bytes_total` и `ml_http_response_size_bytes_total`


src/FuncTest/MetricProvidingBehavior.cs
using System;
using System.Net;
using System.Reflection.Metadata;
using System.Threading.Tasks;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.VisualBasic;
using MyLab.ApiClient;
using MyLab.ApiClient.Test;
using MyLab.HttpMetrics;
using Prometheus;
using TestServer;
using Xunit;
using Xunit.Abstractions;

namespace FuncTest
{
    public class MetricProvidingBehavior : ApiClientTest<Startup, ITestService>
    {
        public MetricProvidingBehavior(ITestOutputHelper output) : base(output)
        {
        }

        [Theory]
        [MemberData(nameof(CreateMetricNames))]
        public async Task ShouldExposeRequestCount(string expectedMetricName, string expectedLe)
        {
            //Arrange
            var le = expectedLe != null ? $",le=\"{expectedLe}\"" : "";
            var expectedMetric = $"{expectedMetricName}{{method=\"POST\",path=\"/api/test/post/xxx/data\",status_code=\"200\"{le}}}";
            await TestCall(srv => srv.Post(0, "data"));

            //Act
            var metrics = await TestCall(srv => srv.GetMetrics());

            //Assert
            Assert.Contains(expectedMetric, metrics.ResponseContent);
        }

        [Fact]
        public async Task ShouldExposeUnhandledExceptionMetric()
        {
            //Arrange
            await TestCall(srv => srv.GetException());

            //Act
            var metrics = await TestCall(srv => srv.GetMetrics());

            //Assert
            Assert.Contains("ml_http_unhandled_exception_total{method=\"GET\",path=\"/api/test/get/exception\",status_code=\"200\"}", metrics.ResponseContent);
        }

        [Fact]
        public async Task ShouldNotHideExceptions()
        {
            //Arrange


            //Act
            var resp = await TestCall(srv => srv.GetException());

            //Assert
            Assert.Equal(HttpStatusCode.InternalServerError, resp.StatusCode);
            Assert.True(resp.IsUnexpectedStatusCode);
        }

        public static object[][]CreateMetricNames()
        {
            return new []
            {
                new object[]{"ml_http_request_count_total", null},

                new object[]{"ml_http_request_content_size_bytes_sum", null},
                new object[]{"ml_http_request_content_size_bytes_count", null},
                new object[]{"ml_http_request_content_size_bytes_bucket", "1024"},
                new object[]{"ml_http_request_content_size_bytes_bucket", "2048"},
                new object[]{"ml_http_request_content_size_bytes_bucket", "5120"},
                new object[]{"ml_http_request_content_size_bytes_bucket", "10240"},
                new object[]{"ml_http_request_content_size_bytes_bucket", "20480"},
                new object[]{"ml_http_request_content_size_bytes_bucket", "51200"},
                new object[]{"ml_http_request_content_size_bytes_bucket", "1048576"},
                new object[]{"ml_http_request_content_size_bytes_bucket", "+Inf"},

                new object[]{"ml_http_response_content_size_bytes_sum", null},
                new object[]{"ml_http_response_content_size_bytes_count", null},
                new object[]{"ml_http_response_content_size_bytes_bucket", "1024"},
                new object[]{"ml_http_response_content_size_bytes_bucket", "2048"},
                new object[]{"ml_http_response_content_size_bytes_bucket", "5120"},
                new object[]{"ml_http_response_content_size_bytes_bucket", "10240"},
                new object[]{"ml_http_response_content_size_bytes_bucket", "20480"},
                new object[]{"ml_http_response_content_size_bytes_bucket", "51200"},
                new object[]{"ml_http_response_content_size_bytes_bucket", "1048576"},
                new object[]{"ml_http_response_content_size_bytes_bucket", "+Inf"},

                new object[]{"ml_http_request_duration_seconds_sum", null},
                new object[]{"ml_http_request_duration_seconds_count", null},

                new object[]{"ml_http_request_duration_seconds_bucket", "0.01"},
                new object[]{"ml_http_request_duration_seconds_bucket", "0.02"},
                new object[]{"ml_http_request_duration_seconds_bucket", "0.05"},
                new object[]{"ml_http_request_duration_seconds_bucket", "0.1"},
                new object[]{"ml_http_request_duration_seconds_bucket", "0.2"},
                new object[]{"ml_http_request_duration_seconds_bucket", "0.5"},
                new object[]{"ml_http_request_duration_seconds_bucket", "1"},
                new object[]{"ml_http_request_duration_seconds_bucket", "2"},
                new object[]{"ml_http_request_duration_seconds_bucket", "5"},
                new object[]{"ml_http_request_duration_seconds_bucket", "10"},
                new object[]{"ml_http_request_duration_seconds_bucket", "20"},
                new object[]{"ml_http_request_duration_seconds_bucket", "30"},
                new object[]{"ml_http_request_duration_seconds_bucket", "+Inf"},
            };
        }
    }

    [Api("")]
    public interface ITestService
    {
        [Post("api/test/post/{id}/data")]
        Task Post([Path]int id, [JsonContent]string data);

        [Get("api/test/get/exception")]
        Task GetException();

        [Get("metrics")]
        Task<string> GetMetrics();
    }
}
